pkg/admission/clusterworkspacetype/admission_test.go

```diff
@@ -0,0 +1,605 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterworkspacetype
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/utils/diff"
+
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+)
+
+func createAttr(ws *tenancyv1alpha1.ClusterWorkspace) admission.Attributes {
+	return admission.NewAttributesRecord(
+		ws,
+		nil,
+		tenancyv1alpha1.Kind("ClusterWorkspace").WithVersion("v1alpha1"),
+		"",
+		"test",
+		tenancyv1alpha1.Resource("clusterworkspaces").WithVersion("v1alpha1"),
+		"",
+		admission.Create,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func updateAttr(ws, old *tenancyv1alpha1.ClusterWorkspace) admission.Attributes {
+	return admission.NewAttributesRecord(
+		ws,
+		old,
+		tenancyv1alpha1.Kind("ClusterWorkspace").WithVersion("v1alpha1"),
+		"",
+		"test",
+		tenancyv1alpha1.Resource("clusterworkspaces").WithVersion("v1alpha1"),
+		"",
+		admission.Update,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func TestAdmit(t *testing.T) {
+	tests := []struct {
+		name        string
+		types       []*tenancyv1alpha1.ClusterWorkspaceType
+		a           admission.Attributes
+		expectedObj runtime.Object
+		wantErr     bool
+	}{
+		{
+			"adds initializers during transition to initializing",
+			[]*tenancyv1alpha1.ClusterWorkspaceType{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "root:org#$#foo",
```

Comment:
Is the name in this format to make the fake lister work? Because normally it would be ClusterName=root:org, Name=foo, right?

Reply:
yes, it's a mock for a multi-workspace lister.